src/main/java/seedu/taassist/logic/commands/result/CommandResult.java

```diff
@@ -0,0 +1,126 @@
+     * Constructs a {@code CommandResult} with the specified {@code feedbackToUser}.
+     */
+    public CommandResult(String feedbackToUser) {
+        requireNonNull(feedbackToUser);
+        this.feedbackToUser = feedbackToUser;
+        uiAction = NO_UI_ACTION;
+        storageAction = NO_STORAGE_ACTION;
+    }
+
+    /**
+     * Constructs a {@code CommandResult} with the specified {@code feedbackToUser} and {@code UiAction}.
+     */
+    public CommandResult(String feedbackToUser, UiAction uiAction) {
+        requireNonNull(feedbackToUser);
+        requireNonNull(uiAction);
+        this.feedbackToUser = feedbackToUser;
+        this.uiAction = uiAction;
+        storageAction = NO_STORAGE_ACTION;
+    }
+
+    /**
+     * Constructs a {@code CommandResult} with the specified {@code feedbackToUser} and {@code StorageAction}.
+     */
+    public CommandResult(String feedbackToUser, StorageAction storageAction) {
+        requireNonNull(feedbackToUser);
+        requireNonNull(storageAction);
+        this.feedbackToUser = feedbackToUser;
+        uiAction = NO_UI_ACTION;
+        this.storageAction = storageAction;
+    }
+
+    public String getFeedbackToUser() {
+        return feedbackToUser;
+    }
+
+    public UiAction getUiAction() {
+        return uiAction;
+    }
+
+    public StorageAction getStorageAction() {
+        return storageAction;
+    }
+
+    public boolean hasUiAction() {
+        return !uiAction.equals(NO_UI_ACTION);
+    }
+
+    public boolean hasStorageAction() {
+        return !storageAction.equals(NO_STORAGE_ACTION);
+    }
+
+    /**
+     * Performs the {@code StorageAction} on the {@code storage} and returns the result of the action.
+     * The {@code StorageAction} must exist.
+     */
+    public CommandResult performStorageAction(Storage storage) throws CommandException {
+        requireNonNull(storage);
+        if (!hasStorageAction()) {
+            throw new CommandException(MESSAGE_NO_STORAGE_ACTION);
+        }
+        StorageActionResult storageActionResult = storageAction.act(storage);
+        return new CommandResult(storageActionResult.combineFeedback(feedbackToUser));
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof CommandResult)) {
+            return false;
+        }
+
+        CommandResult otherCommandResult = (CommandResult) other;
+        return feedbackToUser.equals(otherCommandResult.feedbackToUser)
+                && uiAction.equals(otherCommandResult.uiAction)
+                && storageAction.equals(otherCommandResult.storageAction);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(feedbackToUser, uiAction, storageAction);
+    }
+
+    /**
+     * Represents a {@code StorageAction} that does nothing.
+     */
+    private static class NoStorageAction implements StorageAction {
```

Comment:
This class inside `CommandResult` feels out of place.

Reply:
Let's change the `has` commands to check for null instead so we can leave this out :)

Reply:
I managed to remove studs totally so this will no be an issue now :)